OpenShift Deployment Wizard

dkorn opened the pull request:
This commit introduces the new OpenShift Deployment wizard UI.

To show the Deployment Wizard:
![manageiq create containers provider](https://cloud.githubusercontent.com/assets/6347577/15140642/bdd5cba8-16a4-11e6-918f-bdb4a88b0491.png)

1A General
![deployment_wizard_1](https://cloud.githubusercontent.com/assets/6347577/16222415/892137f4-37a0-11e6-9d32-6973f013215d.png)

1B New VM Settings
![deployment_wizard_2](https://cloud.githubusercontent.com/assets/6347577/16222422/96610b92-37a0-11e6-8eac-4a9f2d943e57.png)

1B Masters & Nodes
![deployment_wizard_3](https://cloud.githubusercontent.com/assets/6347577/16222434/a654fafe-37a0-11e6-97a7-069702afe730.png)

1C CDN Channel
![deployment_wizard_4](https://cloud.githubusercontent.com/assets/6347577/16222442/ad88a47e-37a0-11e6-91d6-36274a892cea.png)

2A Authentication Type 
![deployment_wizard_5](https://cloud.githubusercontent.com/assets/6347577/16222452/bca7e3d4-37a0-11e6-92ba-80000a39b471.png)

2B Authentication Details
![deployment_wizard_7](https://cloud.githubusercontent.com/assets/6347577/16222466/db67c23a-37a0-11e6-998d-3e9194940cbf.png)

2C Settings
![deployment_wizard_12](https://cloud.githubusercontent.com/assets/6347577/16222487/f6541896-37a0-11e6-879b-b1041a5958e8.png)

3A Summary
![deployment_wizard_15](https://cloud.githubusercontent.com/assets/6347577/16222496/0a859c7c-37a1-11e6-8f1f-447eaecea317.png)

3B Progress
![deployment_wizard_16](https://cloud.githubusercontent.com/assets/6347577/16222501/14c39d7e-37a1-11e6-9420-03498a24db5c.png)

dkorn commented:
@miq-bot add_label providers/containers, enhancement, wip, ui

dkorn commented:
@alongoldboim updated

alongoldboim commented:
@dkorn thanks, ill rebase on top of it.

dkorn commented on app/helpers/application_helper/toolbar/ems_containers_center.rb:
```diff
@@ -23,6 +23,12 @@ class ApplicationHelper::Toolbar::EmsContainersCenter < ApplicationHelper::Toolb
+          'icon fa fa-plus fa-lg',
+          t = N_('Deploy a New #{ui_lookup(:table=>"ems_container")}'),
+          t,
+          :html_attributes => {'onClick' => 'showListener()'}),
```
@martinpovolny ping

martinpovolny replied:
@dkorn : how can I reach the toolbar in the UI? cannot find it :-(

martinpovolny replied:
@dkorn : actually I can reach it but it's disabled

martinpovolny commented:
@dkorn : so the changes from #8079 got broken by rebase on master you did. I fixed that by rebasing  #8079 (there's a 2nd place to add the attribute now).

Also I added hack by @himdel to support `data-function=` in the toolbar as discussed on gitter.

serenamarie125 commented:
@dkorn Spoke with @beanh66
Please use the same icon for both menu items ( pficon-add-circle-o ), but make label changes:
Change “Add a New Containers Provider” to “Add Existing Containers Provider"
Change “Deploy a New Containers Provider” to “Create Containers Provider"

Note that at this point, PatternFly is suggesting using the same icon for both Add and Create

@simon3z please be aware of our suggestions above ^^

zeari commented on app/assets/javascripts/application.js:
```diff
@@ -17,6 +17,11 @@
+//= require controllers/deployment/wizard/wizard-directive
+//= require controllers/deployment/wizard/wizard-buttons
+//= require controllers/deployment/wizard/wizard-step-directive
+//= require controllers/deployment/wizard/wizard-substep-directive
```
Cant we put these in directives or `app/assets/javascripts/controllers/deployment/`?

himdel replied:
Either way, please don't clutter application.js with these, since the next 3 lines will include them anyway. (And if you're having problems with module  not being included first, I suggest calling it wizard.js)..

dkorn replied:
@zeari @himdel I removed these 4 lines from ```application.js``` and renamed ```wizard-module.js``` => ```wizard.js``` thanks!

zeari commented on app/assets/javascripts/controllers/deployment/deploy-provider-controller.js:
```diff
@@ -0,0 +1,159 @@
+      } else {
+        $scope.nextButtonTitle = "Next >";
+      }
+    });
```
Might be nicer to have `"Next >"` and the like as constants in another file

dkorn replied:
> Might be nicer to have "Next >" and the like as constants in another file

@zeari thing is, ```$scope.nextButtonTitle``` isn't a constant. In fact, as you can see from the above code, it changes from one step to another.

dkorn commented:
> @dkorn Spoke with @beanh66
> Please use the same icon for both menu items ( pficon-add-circle-o ), but make label changes:
> Change “Add a New Containers Provider” to “Add Existing Containers Provider"
> Change “Deploy a New Containers Provider” to “Create Containers Provider"
> 
> Note that at this point, PatternFly is suggesting using the same icon for both Add and Create
> 
> @simon3z please be aware of our suggestions above ^^

@serenamarie125 @simon3z changed, please approve:
![manageiq create containers provider2](https://cloud.githubusercontent.com/assets/6347577/15140677/e9b295bc-16a4-11e6-8f59-a23db9b538e6.png)

dkorn commented:
@dclarizio @h-kataria this is still a WIP but I wanted to ask, is it a must to convert the .html files to .haml?

himdel commented:
> is it a must to convert the .html files to .haml?

@dkorn it is the currently preferred way to do i18n .. so please do :) .. and there are strings you may want to wrap in `_("...")` in there..

pilhuhn commented on app/helpers/application_helper/toolbar/ems_containers_center.rb:
```diff
@@ -19,10 +19,16 @@ class ApplicationHelper::Toolbar::EmsContainersCenter < ApplicationHelper::Toolb
+          'pficon pficon-add-circle-o fa-lg',
+          t = N_('Create #{ui_lookup(:table=>"ems_container")}'),
+          t,
+          :data => {'function' => 'miqCallAngular', 'function-data' => '{ "name": "showListener", "args": [] }'}),
```
Isn't `showListener` a very / too generic name?
Would it make sense to somehow limit the namespace of this to the container provider so that other providers/MiQ platform could have a function with the same name?

dkorn replied:
> Isn't showListener a very / too generic name?
> Would it make sense to somehow limit the namespace of this to the container provider so that other providers/MiQ platform could have a function with the same name?

@pilhuhn I agree. I can rename this function in my controller.
I'm not really sure how to limit the namespace.

dkorn commented:
> @dkorn it is the currently preferred way to do i18n .. so please do :) .. and there are strings you may want to wrap in _("...") in there..
> 
> (There's a gem (also a (different) online tool) called html2haml which may help .. but it does need some cleanup after..)

Thanks @himdel, .html files now converted to .haml
Also. I added i18n support for all strings

dkorn commented:
@miq-bot remove_label wip

zeari commented on app/assets/javascripts/controllers/container_deployment/deploy-provider-controller.js:
```diff
@@ -0,0 +1,255 @@
+        $http.post(url, {"action" : "create", "resource" :  resource}).success(function (response) {
+          'use strict';
+          $scope.deployInProgress = true;
+          debugger;
```
@dkorn

dkorn replied:
@zeari nice catch, thanks!

zeari commented on app/assets/javascripts/controllers/container_deployment/deploy_provider/deploy-provider-config-auth-settings-controller.js:
```diff
@@ -0,0 +1,196 @@
+      });
+
+      return invalid === undefined;
+    };
```
I think `isValidHtPassword` might be a bit more clear

dkorn replied:
@zeari do you think the `is` prefix makes it clearer? 
I'm not convinced, but if so I'll need to change other boolean vars in my controllers.

zeari commented on app/assets/javascripts/controllers/container_deployment/deploy_provider/deploy-provider-config-settings-controller.js:
```diff
@@ -0,0 +1,65 @@
+          if (!validString($scope.data.nfsMetricsServer) || !validString($scope.data.nfsMetricsPath)) {
+            $scope.configStorageComplete = false;
+          }
+        }
```
Unify some of these `if`s maybe?

zeari commented on app/assets/javascripts/controllers/container_deployment/deploy-provider-controller.js:
```diff
@@ -0,0 +1,255 @@
+      $scope.deploymentDetailsGeneralComplete = false;
+      $scope.deployComplete = false;
+      $scope.deployInProgress = false;
+      $scope.nextButtonTitle = "Next >";
```
Do we want to translate this?

zeari commented on app/assets/javascripts/controllers/container_deployment/deploy_provider/deploy-provider-details-general-controller.js:
```diff
@@ -0,0 +1,45 @@
+            name: provider.provider.name
+          })
+        });
+        if (typeof $scope.existingProviders !== 'undefined' && $scope.existingProviders.length > 0) {
```
just wondering, why is this not angular.isUndefined like the rest?

zeari commented on app/assets/javascripts/controllers/container_deployment/deploy_provider/deploy-provider-master-nodes-controller.js:
```diff
@@ -0,0 +1,605 @@
+
+    $scope.masterCountValid = function(count) {
+      if (count !== 1 && count !== 3 && count !== 5) {
+        $scope.mastersWarning = "The number of Masters must be 1, 3, or 5";
```
translation

zeari commented on app/assets/javascripts/controllers/container_deployment/deploy_provider/deploy-provider-master-nodes-controller.js:
```diff
@@ -0,0 +1,605 @@
+
+    $scope.nodeCountValid = function(count) {
+      if (count < 1) {
+        $scope.nodesWarning = "You must select at least one Node";
```
translation

dkorn replied:
done

zeari commented on app/assets/javascripts/controllers/container_deployment/deploy_provider/deploy-provider-review-summary-controller.js:
```diff
@@ -0,0 +1,63 @@
+        $scope.data.editedInventoryText = $scope.data.inventoryText;
+        //$scope.onInventoryTextChange();
+        $scope.showWaitDialog = false;
+      }, 2000);
```
You should probably delete the commented out part

zeari commented:
Overall I think this PR is good:
1) I think its worth to grep all the strings here and make sure theyre translated.
2) Delete commented out code in various places
2) A super basic JS test that makes sure that the wizard doesnt error out on initialization

dkorn commented:
> Overall I think this PR is good:

thanks for the review :)

> 1) I think its worth to grep all the strings here and make sure theyre translated.

tried that didn't really work.. for example, there are many strings that angular ng module uses.

> 2) Delete commented out code in various places

agreed. in the case of the Inventory/config file edition functionality, we still need to decide what to do.

> 2) A super basic JS test that makes sure that the wizard doesnt error out on initialization

Is there an example of something like that you've seen before? (in manageiq preferred)

dkorn commented:
@himdel @skateman @martinpovolny please review :)

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy_provider/deploy-provider-review-summary-controller.js:
```diff
@@ -0,0 +1,57 @@
+      $scope.showAdvancedSettings = !$scope.showAdvancedSettings;
+    };
+
+    //$scope.onInventoryTextChange = function () {
```
We probably don't want to introduce huge commented-out blocks of code in a new file..  Please remove..

dkorn replied:
This will be added back in the future (next release probably), for now - removed. thanks.

himdel commented on app/assets/javascripts/controllers/container_deployment/dialogs.module.js:
```diff
@@ -0,0 +1,8 @@
+ *   Dialogs module for MiQ.
+ *
+ */
+angular.module('miq.dialogs', []);
```
I'm a bit concerned about this file's location .. or rather, the only file that defines stuff in this module is `container_deployment/modal_dialog/modal-dialog-directive.js` .. I mean, this should either be part of that file and not be called `miq.dialogs` but rather something more specific, or if it is called `miq.dialogs` and meant to contain all the miq dialogs, then it makes no sense to put it in `container_deployment`..

dkorn replied:
> I'm a bit concerned about this file's location .. or rather, the only file that defines stuff in this module is container_deployment/modal_dialog/modal-dialog-directive.js .. I mean, this should either be part of that file and not be called miq.dialogs but rather something more specific,

I tried to maintain consistency with the container dashboard modules and directive formats. for example you have `card-module.js` and it's directive `heatmaps-card-directive.js`. I don't mind changing the module name but we probably be consistent.

> or if it is called miq.dialogs and meant to contain all the miq dialogs, then it makes no sense to put it in container_deployment.. (but yeah, that's a detail at best..:) )

that wasn't my intention :)

himdel replied:
Definitely agreed on consistency, even though it may be hard to find any now :). I'll leave this up to you, if you can figure out a more fitting name, feel free, if not, introducing some consistency with modules is on my list anyway :)

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy_provider/deploy-provider-review-controller.js:
```diff
@@ -0,0 +1,89 @@
+      return selectedTemplate ? selectedTemplate.name : "None";
+    };
+
+    $scope.getAuthenticationType = function () {
```
missing translations..

dkorn replied:
you meant the `"None"` string, right?

himdel replied:
I mean all the returned strings in getAuthenticationType, since they are shown to the user.. I agree that translating `Google` may make less sense than translating `Request Header`, but that should be up to the translators, not us..

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy_provider/deploy-provider-review-controller.js:
```diff
@@ -0,0 +1,89 @@
+    };
+
+    $scope.getConfigServerType = function() {
+      if ($scope.data.serverConfigType == 'standardNFS') {
```
same here..

dkorn replied:
I added translation only for the `"None"` string, since the others are authentication terms

himdel replied:
I don't think so.. maybe not many translation will translate `GitHub`, but `Request Header` or `Allow All`  are definitely candidates for translation. And so is `Standard NFS server` vs `Integrated NFS server`.

Maybe more to the point, the decision should be on the translators here.

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy-provider-controller.js:
```diff
@@ -0,0 +1,254 @@
+      $timeout(function () {
+        var url = '/api/container_deployments';
+        var resource = create_deployment_resource();
+        $http.post(url, {"action" : "create", "resource" :  resource}).success(function (response) {
```
Don't use `$http` to talk to the API, `$http` doesn't send the right token.. use the `API` service instead..

dkorn replied:
done, thanks.

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy-provider-controller.js:
```diff
@@ -0,0 +1,254 @@
+
+    $scope.ready = false;
+    var url = '/api/container_deployments/container_deployment_data';
+    $http.get(url).success(function (response) {
```
same here, should be `API.get(url).then(...)`

dkorn replied:
done, thanks.

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy-provider-controller.js:
```diff
@@ -0,0 +1,254 @@
+          $scope.deployComplete = false;
+          $scope.deploySuccess = false;
+          $scope.deployFailed = false;
+          $scope.deployFailureMessage = "An unknown error has occurred.";
```
missing translation

dkorn replied:
done, thanks.

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy_provider/deploy-provider-config-auth-settings-controller.js:
```diff
@@ -0,0 +1,196 @@
+        firstShow = false;
+      }
+      $scope.validateForm();
+      $timeout(function() {
```
pretty sure we have an `auto-focus` directive for this...

himdel commented:
@dkorn I'd really appreciate if you replaced that `$document` abuse with some autofocus directive in all those places... (using `$document` for file input manipulation in `browseKeyFile` seems OK) .. but it *can* be done in a separate PR..

simon3z commented:
> > @simon3z please be aware of our suggestions above ^^
> 
> @serenamarie125 @simon3z changed, please approve:

@serenamarie125 @dkorn I'd prefer something that makes the user understand that he's going to deploy OpenShift; “Create Containers Provider" doesn't seem to recall that.

Maybe "Deploy a New Container Provider"?
(And leave the other one consistent with the other providers for VMs).

dkorn commented:
> @dkorn I'd really appreciate if you replaced that $document abuse with some autofocus directive in all those places... (using $document for file input manipulation in browseKeyFile seems OK) .. but it can be done in a separate PR..
> 
> Alternatively, if you really need to do this dynamically, at least wrap the pattern in a function, so that we get focus('foo') instead of $timeout + $document + .focus() everywhrere..

I added the `dynamicAutoFocus` function to `miq_service.js` and dropped the `$timeout` and `$document` from all the controllers, except for the browseKeyFile functionality.
Let me know what you think of it.

dkorn commented:
@h-kataria please review

h-kataria commented:
@martinpovolny @himdel is this good to go? @dkorn can you take a look at failing spec test.

dkorn commented:
@h-kataria the fail is in `spec/requests/api/entrypoint_spec.rb`, I don't think it's connected to this PR

dkorn commented:
@jeff-phillips-18 please review :)

jeff-phillips-18 commented:
Looks really good to me, almost exactly how I would have done it (minus the haml) ;)

dclarizio commented:
@himdel @martinpovolny you guys agree this is ready?

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy-provider-controller.js:
```diff
@@ -0,0 +1,254 @@
+        }
+      };
+
+      $timeout(function() {
```
This feels wrong, why the 5 second wait?

dkorn replied:
because it takes time for the api request to return with the needed info and then load it to the relevant variable and fields.
if this is removed on most tries the substeps don't load fast enough.

himdel replied:
So why not wait for the promise of that api request, instead of hardcoding 5 seconds. There's no arbitrary number that will always be enough..

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy-provider-controller.js:
```diff
@@ -0,0 +1,254 @@
+      var name = "";
+      var publicName = "";
+      $scope.nodeData.allNodes.forEach(function(item) {
+        if ($scope.data.provisionOn == 'existingVms') {
```
Is it intentional that for `existingVms` you'll use the publicName of the previous node, and for `noProvider`, you'll reuse the id of the previous node?

Because if not, `var id, name, publicName` should really happen only within that forEach..

dkorn replied:
> Is it intentional that for existingVms you'll use the publicName of the previous node, and for noProvider, you'll reuse the id of the previous node?
> 
> Because if not, var id, name, publicName should really happen only within that forEach..
> 
> (And if it is, it warrants a comment explaining why)

moved inside the forEach, thanks.

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy-provider-controller.js:
```diff
@@ -0,0 +1,254 @@
+          $scope.deployFailed = false;
+          $scope.deployFailureMessage = __("An unknown error has occurred.");
+        }
+      }, 5000);
```
Another 5s timeout..

dkorn replied:
without it there's a stall till the spinner and messages appear

himdel replied:
Aaaah... you need to inject `API` into the controller...

Right now, you're using the global `API` variable, that provides API access to non-angular code and is not synced with the angular's `$digest` cycle. If you inject `API`, it will use the angular service, whcih is exactly the same, except triggers a `$digest` immediately after the response..

himdel commented:
@dkorn also, would you move those templats under `static` to a subdirectory? Something like `static/deployment_wizard/` maybe?

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy_provider/deploy-provider-review-progress-controller.js:
```diff
@@ -0,0 +1,13 @@
+  function($rootScope, $scope) {
+    'use strict';
+
+    var firstShow = true;
```
this seems to do nothing at all..?

dkorn replied:
`firstShow` in `onShow` function is used all across the wizard controllers, to support moving between steps. we plan to add functionality to this substep , in a different PR.
I removed this function for now, thanks.

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy_provider/deploy-provider-review-summary-controller.js:
```diff
@@ -0,0 +1,36 @@
+      // $scope.data.review_inventory = true;
+      $scope.showWaitDialog = true;
+      $timeout(function() {
+        $scope.data.inventoryText = "Text of the Inventory goes here";
```
i18n

dkorn replied:
currently, we are keeping it an empty string so no i18n is required.

himdel commented on app/assets/javascripts/controllers/container_deployment/wizard/wizard-review-page-directive.js:
```diff
@@ -0,0 +1,40 @@
+      };
+      $scope.getReviewSubSteps = function (reviewStep) {
+        var substeps = reviewStep.getReviewSteps();
+        console.log("Step " + reviewStep.stepTitle + " has " + substeps.length + " steps");
```
Guessing this was forgotten here?

dkorn replied:
right, thanks.

himdel commented on app/assets/javascripts/controllers/container_deployment/wizard/wizard-step-directive.js:
```diff
@@ -0,0 +1,397 @@
+      if (angular.isUndefined($scope.stepPriority)) {
+        $scope.stepPriority = 999;
+      } else {
+        $scope.stepPriority = parseInt($scope.stepPriority);
```
`parseInt` without an explicit base is ..problematic, please add `, 10`

dkorn replied:
done, thanks.

himdel commented on app/views/static/deploy-provider-auth-settings-review.html.haml:
```diff
@@ -0,0 +1,83 @@
+  %form.form-horizontal{"ng-if" => "data.authentication.mode == 'htPassword'"}
+    .wizard-miq-review-inline{"ng-repeat" => "user in data.authentication.htPassword.users track by $index"}
+      .wizard-miq-review-item
+        %span.wizard-miq-review-item-label Username:
```
missing translations for the labels

dkorn replied:
done, thanks.

himdel commented on app/views/static/deploy-provider-details-cdn-review.html.haml:
```diff
@@ -0,0 +1,18 @@
+        = _("CDN Channel:")
+      %span.wizard-miq-review-item-value {{data.cdnEnabled ? "Enabled" : "Disabled"}}
+    .wizard-miq-review-item{"ng-if" => "data.cdnEnabled"}
+      %span.wizard-miq-review-item-label Username:
```
same here..

dkorn replied:
done, thanks.

himdel commented on app/views/static/deploy-provider-details-existing-vms.html.haml:
```diff
@@ -0,0 +1,137 @@
+              %a.select-all-roles.btn-link{"ng-click" => "toolbarMenuSelectAll()"} Select All
+              %label.checkbox
+                %input{"ng-model" => "toolbarMenu.master", :role => "presentation", :type => "checkbox"}
+                  Master
```
Bad indent here, `input` is not a pair-element and should not contain anything inside..

Also, missing translation

martinpovolny replied:
more missing i18n below

dkorn replied:
@himdel not sure I understand what you meant here. these are checkboxes, therefore should be input, right? 
added the translation (requires a new line), thanks.

himdel replied:
Yes, defeinitely should be an `input`, and yes, new line is good .. but delete the two spaces in front of that new line :)

My problem was that

```haml
%input
  Text
```

means

```html
<input>Text</input>
```

But input has no inside in html, it should be

```html
<input />
Text
```

dkorn replied:
@himdel got it, fixed, thanks

himdel commented on app/views/static/deploy-provider-details-existing-vms.html.haml:
```diff
@@ -0,0 +1,137 @@
+            %td
+              .role-column
+                %span.label.label-primary{"ng-if" => "node.master"}
+                  Master
```
i18n

dkorn replied:
done, thanks.

himdel commented on app/views/static/deploy-provider-details-existing-vms.html.haml:
```diff
@@ -0,0 +1,137 @@
+    .table-footer
+      .badge{:tooltip => "{{mastersWarning}}", "tooltip-placement" => "right"}
+        %span.pficon.pficon-warning-triangle-o{"ng-if" => "mastersWarning && mastersWarning != ''"}
+        %span{"ng-if" => "mastersCount == 1"} {{mastersCount}} Master
```
i18n here too..

martinpovolny replied:
and below...

dkorn replied:
done, thanks.

himdel commented on app/views/static/deploy-provider-details-no-provider.html.haml:
```diff
@@ -0,0 +1,180 @@
+              %a.select-all-roles.btn-link{"ng-click" => "toolbarMenuSelectAll()"} Select All
+              %label.checkbox
+                %input{"ng-model" => "toolbarMenu.master", :role => "presentation", :type => "checkbox"}
+                  Master
```
Same input problem here..

dkorn replied:
added i18n, not sure what's the input problem.

himdel commented on app/views/static/deploy-provider-master-nodes-review-create-vms.html.haml:
```diff
@@ -0,0 +1,22 @@
+  %form.wizard-miq-review-subform
+  %form.wizard-miq-review-subform
+    .wizard-miq-review-item
+      %span.wizard-miq-review-item-label Masters Base Name:
```
i18n

dkorn replied:
done, thanks.

himdel commented on app/views/static/deploy-provider-master-nodes-review-tables.html.haml:
```diff
@@ -0,0 +1,8 @@
+.wizard-miq-review-item
+  .wizard-miq-review-item-field {{data.masters.length}} {{data.masters.length === 1 ? 'Master' : 'Masters'}}
```
i18n

dkorn replied:
done, thanks.

himdel commented on app/views/static/deploy-provider-review-progress.html.haml:
```diff
@@ -0,0 +1,23 @@
+  "step-priority" => "1", "step-title" => "Progress", "wz-disabled" => "false"}
+    .blank-slate-pf{"ng-if" => "deployInProgress"}
+      .spinner.spinner-lg
+      %h1 Deployment in Progress
```
i18n :)

dkorn replied:
done, thanks.

himdel commented on app/views/static/deploy-provider-review-progress.html.haml:
```diff
@@ -0,0 +1,23 @@
+    .blank-slate-pf{"ng-if" => "deployComplete"}
+      .alert.alert-success{"ng-if" => "deploySuccess"}
+        %span.pficon.pficon-ok
+        Deployment
```
This cannot be split because of i18n, need to do `= _("Deployment %{name} was sucessful") % {:name => "{{data.providerName}}""}

dkorn replied:
done, thanks.

himdel commented on app/views/static/deploy-provider-review-progress.html.haml:
```diff
@@ -0,0 +1,23 @@
+        was successful.
+      .alert.alert-danger{"ng-if" => "deployFailed"}
+        %span.pficon.pficon-error-circle-o
+        Failed to deploy
```
similar here..

dkorn replied:
done, thanks.

himdel commented on app/views/static/deploy-provider-review-summary.html.haml:
```diff
@@ -0,0 +1,23 @@
+            %form
+              %textarea.miq-review-edit-text{"ng-change" => "onInventoryTextChange()",
+              "ng-model" => "data.editedInventoryText", :readonly => "", :rows => "10", :type => "text"}
+              / <div class="modal-footer miq-review-button-panel">
```
Please don't introduce commented-out code..

dkorn replied:
done, thanks.

himdel commented on app/views/static/wizard.html:
```diff
@@ -0,0 +1,31 @@
+    <div ng-if="!wizardReady" class="wizard-miq-step-content" ng-style="contentStyle">
+      <div class="blank-slate-pf">
+        <div class="spinner spinner-lg"></div>
+        <h1>Initialization in progress...</h1>
```
i18n .. so will probably need haml..

dkorn replied:
done, thanks.

dkorn commented:
@dclarizio this should be backported to darga, so we'll need darga/yes

martinpovolny commented on app/views/static/deploy-provider-details-existing-vms.html.haml:
```diff
@@ -0,0 +1,137 @@
+                  %a
+                    %span.fa.fa-close{"aria-hidden" => "true", "ng-click" => "removeDns(node)"}
+                %span.label.label-primary{"ng-if" => "node.etcd"}
+                  Etcd
```
i18n here and above

dkorn replied:
> i18n here and above

@martinpovolny my initial thought was not to use i18n on the roles (Master, Node, Etcd, Load Balancer, Storage, Infra, DNS) since they are terms.
I've seen both you and @himdel commenting on it so I'll add that.

martinpovolny commented on db/fixtures/miq_product_features.yml:
```diff
@@ -3275,6 +3275,10 @@
+      :description: Deploy a New Containers Provider
+      :feature_type: admin
+      :identifier: ems_container_deployment
 
```
you might also need to add the feature to `db/fixtures/miq_user_roles.yml`

dkorn replied:
added, thanks.

martinpovolny commented:
Pretty significant piece of work :+1:

dkorn commented:
> EDIT: or deploy_provider since that's what you're using for JS (even though that's a bit confusing in the context of angular (provider != provider))

dkorn commented:
> @dkorn also, would you move those templats under static to a subdirectory? Something like static/deployment_wizard/ maybe?
> 
> EDIT: or deploy_provider since that's what you're using for JS (even though that's a bit confusing in the context of angular (provider != provider))

@himdel I'll move them to a subdirectory. how about: static/deploy_containers_provider ?

himdel commented:
> @himdel I'll move them to a subdirectory. how about: static/deploy_containers_provider ?

sounds good :)

dkorn commented:
@himdel IINM I handled all your review comments. let me know if something else is needes.

dkorn commented:
@himdel @dclarizio @h-kataria bumping @martinpovolny bumping this PR again. 
Unless I missed something, I believe I fixed all the review comments so this is ready for merge.

Worth mentioning - we currently decided to hide the wizard from the UI, and plan to change it during September 
@simon3z @alongoldboim

himdel commented on app/assets/stylesheets/wizard.scss:
```diff
@@ -0,0 +1,946 @@
+    text-align: center;
+
+    .btn[disabled] {
+      pointer-events: none;
```
Note that this does not prevent the user from activating the button by keyword, and does break any `cursor` style and tooltips. Sure about it?

himdel commented on app/assets/javascripts/controllers/container_deployment/deploy-provider-controller.js:
```diff
@@ -0,0 +1,247 @@
+        $scope.deployFailed = response.error !== undefined;
+        if (response.error) {
+          if (response.error.message) {
+            $scope.deployFailureMessage = __(response.error.message);
```
that `__` is *probably* broken here.. either the message already comes translated, and you don't want to translate it again, or it comes in english, but unless it's marked with `N_()` on the server side, it won't get into the catalog..

dkorn replied:
@himdel so I should just drop the `__` here, right?

himdel replied:
Sure, makes sense :) Unless there's like 2 possible error messages in which case a `switch` may make sense..

himdel commented:
All the text in `static/deploy_containers_provider/deploy-provider-master-nodes-review-tables.html.haml` is currently broken..

`_("%{master}") % {:master => "{{data.masters.length === 1 ? 'Master' : 'Masters'}}"}` means that `Master` or `Masters` will never be translated

`_(%{node})` means `_("node")` and never even touches the stuff after `%..` (missing quotes, but then we're back to the master problem..)

And DNS and Storage should probably get similar logic..


The right way to fix it would be using `n_()` but we don't have that yet in JS.. for now, at least something like `data.masters.length === 1 ? __("Master") : __("Masters")` (without that left side)?

himdel commented:
Apart from those 3, I'm not seeing any other problems..

dclarizio commented:
@dkorn will await replies to @himdel latest comments prior to merging.  Can you also clarify this:
>Worth mentioning - we currently decided to hide the wizard from the UI, and plan to change it during September 

Does  this mean it's `turned off` somehow and you will revisit, update, and re-enable it later?

cben commented:
The closest we got is this:
```
{{data.masters.length === 1 ? __('1 Master') : sprintf(__('%s Masters'), data.masters.length)}}
```
The expression works in JS console but the angular template silently prints nothing :-(
Angular makes it hard to debug: https://stackoverflow.com/questions/32764895/how-to-print-to-console-log-from-inside-angular-js-inline-templates-script-tag

IIUC the problem is that templates are evalated in very sandboxed context =>`sprintf` is not available, would have to be explicitly exposed (in contoller, or somewhere more centralized?).  I think `{{__('foo')}}` did work, though not 100% sure now.

Quick search suggests we currently have zero `__` calls inside angular templates, so this is new ground.
Please advise.

himdel commented:
@cben I think the problem may be in that `%s` expects a string but `data.masters.length` is a number.. try `%d` instead..

Basically, if `__` is available, then so is `sprintf`, both are properties on `window`..

dclarizio commented:
@cben will you be addressing @himdel's [comment](https://github.com/ManageIQ/manageiq/pull/7620#issuecomment-241273631)?

dkorn commented:
@dclarizio I will, thanks.

dkorn commented:
@himdel No %d doesn't help.  Neither `sprintf` nor `__` are available in angular {{templates}}.
this is the solution we (@cben) found:
`= _("{{data.masters.length}} {{data.masters.length === 1 ? 'Master' : 'Masters'}}")`
which is an ugly string to translate but does give translator control over word order.

> And DNS and Storage should probably get similar logic..

there can only be one of each so no need of plural forms.



> Does this mean it's turned off somehow and you will revisit, update, and re-enable it later?

@dclarizio yes, currently till all the backend dependencies are merged, the wizard is hidden by default. It is possible to display the wizard by going to configurations -> advanced -> :product: `:container_deployment_wizard: true => true`

himdel commented:
> `"{{data.masters.length}} {{data.masters.length === 1 ? 'Master' : 'Masters'}}"`

@dkorn we can't seriously give translators that kind of a string..

As I see it, there are 3 options to fix this:

   * `$rootScope.__ = __; $rootScope.sprintf = sprintf;` (during `app.run`)- and voila, you can use both in templates
   * `$scope.$watch('data.masters.length', function(len) { $scope.text = (len > 1) ? __("Master") : __("Masters") })` - in the controller, and then just `{{text}}` (but with a better name)
   * use ng-show/hide to show the right translation in the template..   

```
    %div{"ng-show" => "data.masters.length === 1"}
      = _('Master')
    %div{"ng-hide" => "data.masters.length === 1"}
      = _('Masters')
```

I think the third one is the clearest but feel free to choose :)

dkorn commented:
@himdel I chose option 3 as you recommended, just added the number of each role nodes:
```
%div{"ng-show" => "data.masters.length === 1"}
  = "{{data.masters.length}} " + _('Master')
%div{"ng-hide" => "data.masters.length === 1"}
  = "{{data.masters.length}} " + _('Masters')
```

cben commented:
@martinpovolny `n()` is Ruby, right?  That will not work as the number is only known in the browser.
javascript `n_()` requires 1st or 2nd option to be able to call it.
[I have no opinion among the 3 options.]

dkorn commented:
@himdel bumping this again, I've made the change you asked (option 3)

himdel commented:
Looks mergable, thanks, merging!

I will be assigning you to an issue listing some of the remaining problems I've found though :).

himdel commented:
To enable in the UI, go to the server config, advanced, `:product` section, and set `container_deployment_wizard` to true..

dkorn commented:
> I will be assigning you to an issue listing some of the remaining problems I've found though :).

Great, Thanks!

himdel commented:
https://github.com/ManageIQ/manageiq/issues/11068 :)
